### PR TITLE
feat: customizable management command to generate local enterprise data

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
How to use this work:

Upon invoking the command, supply a model param (`--model`) that is a complete path to a model that has a corresponding test factory ie

`./manage.py lms manufacture_data --model enterprise.models.EnterpriseCustomer`

This will generate an enterprise customer record with place holder values according to the [test factory](https://github.com/openedx/edx-enterprise/blob/master/test_utils/factories.py#L89)

```
'EnterpriseCustomer' fields: {}
```

We can also provide customizations to the record being generated, ie

`./manage.py lms manufacture_data --model enterprise.models.EnterpriseCustomer --name "FRED"`

which will produce the customized record:

```
'EnterpriseCustomer' fields: {'name': 'FRED'}
```

we can also supply parent model/subfactory customizations as well: 

`./manage.py lms manufacture_data --model enterprise.models.EnterpriseCustomerCatalog --enterprise_customer__site__name "Fred" --enterprise_catalog_query__title "JOE SHMO" --title "who?"` will result in
```
'EnterpriseCustomerCatalog' fields: {'title': 'who?'}
	'EnterpriseCustomer' fields: {}
		'Site' fields: {'name': 'Fred'}
	'EnterpriseCatalogQuery' fields: {'title': 'JOE SHMO'}
```

Note the non subclass customization `--title "who?"` is applied to the specified model `EnterpriseCustomerCatalog`

But say we want to supply an existing record as a FK to our object:

`./manage.py lms manufacture_data --model enterprise.models.EnterpriseCustomerUser --enterprise_customer 994599e6-3787-48ba-a2d1-42d1bdf6c46e`

we get:

```
'EnterpriseCustomerUser' fields: {}
	'EnterpriseCustomer' PK: 994599e6-3787-48ba-a2d1-42d1bdf6c46e
```

or we can do something like:

`./manage.py lms manufacture_data --model enterprise.models.EnterpriseCustomerUser --enterprise_customer__site 9 --enterprise_customer__name "joe"`

which would yield:

```
'EnterpriseCustomerUser' fields: {}
	'EnterpriseCustomer' fields: {'name': 'joe'}
		'Site' PK: 9
```

but if you try and get something that doesn't exist...

`./manage.py lms manufacture_data --model enterprise.models.EnterpriseCustomerUser --enterprise_customer <SOMETHING BAD>`

we'd get:

```
CommandError: Provided FK value: <SOMETHING BAD> does not exist on EnterpriseCustomer
```

another limitation of this script is that it can only fetch or customize, you cannot customize a specified, existing FK...

` ./manage.py lms manufacture_data --model enterprise.models.EnterpriseCustomerUser --enterprise_customer__site__name "fred" --enterprise_customer 994599e6-3787-48ba-a2d1-42d1bdf6c46e`

would yield a 

`CommandError: This script does not support customizing provided existing objects`

https://2u-internal.atlassian.net/browse/ENT-7403

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/openedx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/openedx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/openedx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/openedx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/openedx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: enterprise/management/commands/manufacture_factory_data.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        factories_list = all_subclasses(factory.django.DjangoModelFactory)
+        # Find the factory that matches the provided model
+        for potential_factory in factories_list:
+            factory_model = potential_factory._meta.model
+            if f"{factory_model.__module__}.{factory_model.__name__}" == provided_model:
+                packed_factory_kwargs = {}
+                paired_attributes = pairwise(self.custom_attributes)
+                # Confirm that the provided custom attributes are valid fields on the provided model
+                for field, value in paired_attributes:
+                    striped_field = field.strip("--")
```
Comment (marlonkeating) [inline enterprise/management/commands/manufacture_factory_data.py:129]:
```suggestion
                    stripped_field = field.strip("--")
```

**Review by marlonkeating — CHANGES_REQUESTED** [review]

File: enterprise/management/commands/manufacture_factory_data.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    return string.replace("_", " ").title().replace(" ", "")
+
+
+def pairwise(iterable):
+    """
+    Convert a list into a list of tuples of adjacent elements.
+    s -> [ (s0, s1), (s2, s3), (s4, s5), ... ]
+    """
+    a = iter(iterable)
+    return zip(a, a)
```
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply enterprise/management/commands/manufacture_factory_data.py:42]: I'm sure I tested this... how was it working locally..

File: enterprise/management/commands/manufacture_factory_data.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+# We have to import the enterprise test factories to ensure it's loaded and found by __subclasses__
+from test_utils import factories as _
+
+# To ensure factories outside of the enterprise package are loaded and found by the script,
+# add any additionally desired factories as an import to this file. Make sure to catch the ImportError
+# incase other consumers of the command do not have the same factories installed.
+# For example:
+try:
+    import common.djangoapps.student.test.factories as _
```
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply enterprise/management/commands/manufacture_factory_data.py:22]: yes... technically should be two different, although I think functionally it doesn't matter as we just need the modules imported for the subclassing to work.

File: enterprise/management/commands/manufacture_factory_data.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    def handle(self, *args, **options):
+        """
+        Entry point for management command execution.
+        """
+        if not options.get('model'):
+            log.error("Did not receive a model")
+            raise CommandError("Did not receive a model")
+
+        path_of_model = options.get('model').split(".")
+        provided_model = '.'.join(path_of_model[:-1]) + '.' + convert_to_pascal(path_of_model[-1])
```
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply enterprise/management/commands/manufacture_factory_data.py:117]: being extra careful + it felt cleaner as everything else in the command was snake case so it gives flexibility to the user

[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply enterprise/management/commands/manufacture_factory_data.py:117]:
```
$ ./manage.py manufacture_factory_data --model enterprise.models.enterprise_customer \
            --name "Test Enterprise" --slug "test-enterprise"
```
vs 
```
$ ./manage.py manufacture_factory_data --model enterprise.models.EnterpriseCustomer \
            --name "Test Enterprise" --slug "test-enterprise"
``` 
the point is just to have both work ¯\_(ツ)_/¯

File: enterprise/management/commands/manufacture_factory_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply enterprise/management/commands/manufacture_factory_data.py:42]: yeah confirmed locally that this works

File: enterprise/management/commands/manufacture_factory_data.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    def run_from_argv(self, argv):
+        """
+        Re-implemented from https://github.com/django/django/blob/main/django/core/management/base.py#L395 in order to
+        support individual field customization.
+
+        Uses ``parse_known_args`` instead of ``parse_args`` to not throw an error when encountering unknown arguments
+
+        https://docs.python.org/3.8/library/argparse.html#argparse.ArgumentParser.parse_known_args
+        """
+        self._called_from_command_line = True
```
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply enterprise/management/commands/manufacture_factory_data.py:79]: ok I got something working were we just insert into the options object/namespace

File: enterprise/management/commands/manufacture_factory_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (marlonkeating) [reply enterprise/management/commands/manufacture_factory_data.py:42]: Double-checked, and confirmed your current method works.  I missed that you're using an iterator over a single list, which zip will keep iterating through.

File: enterprise/management/commands/manufacture_factory_data.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (marlonkeating) [reply enterprise/management/commands/manufacture_factory_data.py:79]: I'm more concerned about `self._called_from_command_line`.  My understanding of python conventions is that by setting properties that begin with an underscore, we're relying on an internal implementation detail of Command that is subject to change.

**Comment by adamstankiewicz** [comment]
@alex-sheehan-edx [curious] Just seeking clarification on the approach here: how does this new command relate to the existing `seed_enterprise_devstack_data` command? Is the intent to deprecate that existing one in favor of this new one? It might be confusing to have 2 commands documented doing seemingly somewhat similar things.

**Comment by alex-sheehan-edx** [comment]
> 



> Member

The hope is to deprecate the old one! But there are also some use cases for this that I think apply outside of that script that I will try and highlight 😄

**Comment by alex-sheehan-edx** [comment]
> @alex-sheehan-edx [curious] Just seeking clarification on the approach here: how does this new command relate to the existing `seed_enterprise_devstack_data` command? Is the intent to deprecate that existing one in favor of this new one? It might be confusing to have 2 commands documented doing seemingly somewhat similar things.

@adamstankiewicz The hope is to deprecate the old one! But there are also some use cases for this that I think apply outside of that script that I will try and highlight 😄

**Comment by johnnagro** [comment]
> @alex-sheehan-edx [curious] Just seeking clarification on the approach here: how does this new command relate to the existing `seed_enterprise_devstack_data` command? Is the intent to deprecate that existing one in favor of this new one? It might be confusing to have 2 commands documented doing seemingly somewhat similar things.

@adamstankiewicz imho we may keep the seed data script and/or improve it with what we've learned here. the idea with this command - following the unix philosophy - is to make a simple, dedicated provisioning utility that can be used stand-alone or used as part of a script. this should allow us to more easily make more seed data scripts tailored to the specific needs of a feature epic.

**Review by johnnagro — APPROVED** [review]
looks great so far. would like to see some example scripts which utilize it.

File: enterprise/management/commands/manufacture_factory_data.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (marlonkeating) [reply enterprise/management/commands/manufacture_factory_data.py:243]: In reading up a bit more on python conventions, I don't see a strong consensus on whether subclasses should be using single-underscored variables from parent classes.  There's always the risk of a future version of BaseCommand changing what they deem to be an implementation detail and we only find out by this breaking on our end, but I don't see a way around that at present (without copying the whole BaseCommand class).

File: requirements/common_constraints.txt [hunk 6]
```diff
@@ -1,5 +1,7 @@
 
 
+
```
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply requirements/common_constraints.txt:3]: uhm no, what's weird is that I'm pretty sure `make upgrade` did this 🤔

File: tests/test_enterprise/management/test_manufacture_data.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    BaseCommand,
+    CommandError,
+)
+from django.test import TestCase
+
+from enterprise import models
+from django.core.management.base import CommandError
+
+
+def call_command(command_name, *args, **options):
```
Comment (alex-sheehan-edx) [inline tests/test_enterprise/management/test_manufacture_data.py:25]: I had to redefine call_command to allow us to use unrecognized args

**Review by marlonkeating — APPROVED** [review]

**Pull request merged**